### ISSUE TITLE
Should we be concerned about garbage collection when using destructuring syntax?

### ISSUE THREAD

**Issue by loopmode** [parro-it/amaj#4]
Consider these two functions:

Using destructuring:
```
function createUser({name, age, email}) {
  console.log({name, age, email});
  // ...
}
```

Regular:
```
function createUser(name, age, email) {
  console.log(name, age, email);
  // ...
}
```
They do essentially the same.

Maybe my question is obsolete once destructuring is a native feature and doesn't need to be transpiled.
However, when using a transpiler like Babel, the first example ends up like this:

```
function createUser(_ref) {
  var name = _ref.name,
      age = _ref.age,
      email = _ref.email;

  console.log({ name: name, age: age, email: email });
  // ...
}
```

So, we implicitly create temporary objects that that hold our values.

I do this a lot, and many times for reasons that are not the primary use case for this syntax.
In a case where you have many potentially optional arguments, it makes perfect sense to use the syntax, as the callsite doesn't have to know the order of arguments. Also, I've seen Java code (for example) that hurts my eyes, where a constructor takes 10 optional arguments, and if you need to pass the first and the last one, you'll pass 8x `null` between them... terrible.

So, I also use it for aesthetical reasons, but moreover for cosistency reasons: I have entire APIs that work with destructured objects as arguments, even if it's just two or three values. And I'm happy with - it's "just web apps", "just forms and stuff" and I do not care about performance impacts.

But. :)
Let's say you're building a high-performance mechanism, like a 3D app, or a serverside application that has thousands of operations per second, and calls dozens or hundreds such functions in a single stack.

I remember doing a 3D game in Flash back in the day that occasionally would have a hickup and framerate drops from 50fps to 20fps - every time the GC kicked in and collected all the stale event objects (hundreds and thousands MouseMove, EnterFrame and custom event objects)

Libraries like react already use pooling mechanisms for event objects, I guess one reason being to avoid garbage collection.

So now here we are, in the age of ES2016+, and we have the cool new destructuring syntax.

1) Should we be concerned about negative impacts of garbage collection when using destructuring syntax, or should we focus on the comforts we gain?

2) Even when we don't transpile and have native support (e.g. Node.js or Electron apps with definite support) - is this problem out of the world? Or do we just ditch it out of the Javascript layer and into the C layer below?
Of course it depends on the JS engine, but what do we know about e.g. V8 handling this? 
Do we create much extra work for cleaning up our "garbage" inside V8? (I guess, while there is no GC in C++, there still needs to be some cleanup/allocation management)

**Comment by parro-it** [comment]
Thank you, this is a really interesting question.

First of all, I would to like to do some (maybe pedant) clarifications
on your code example.
When you wrote this snippet of babelified code:

```js
function createUser(_ref) {
  var name = _ref.name,
      age = _ref.age,
      email = _ref.email;

  console.log({ name: name, age: age, email: email });
  // ...
}
```

> So, we implicitly create temporary objects that that hold our values.

You seem to mean that this code snippet create new objects that has to be later GC'ed by the JS engine:

```js
var name = _ref.name,
      age = _ref.age,
      email = _ref.email;
```

But this code only create new variables that refers of existing objects. This
should be really fast on every JS engine.

The `console.log` line, instead creates a new object, that has to be garbage collected later, thus slowing your program:

```js
  console.log({ name: name, age: age, email: email });
```

but you probably don't keep that line in production code.

Also, if you don't already have all the arguments stored in an object
at call site, you probably will have to create a new object there:

```js
createUser({name:'Andrea', age:42});	// you are creating a new object here.
```

Anyway, about the actual questions:

> 1) Should we be concerned about negative impacts of garbage collection when using destructuring syntax, or should we focus on the comforts we gain?

If you create lot of objects, you should always consider that garbage collection could have a noticeable negative impact on performance. You could have noticeable performance gains by just reducing the number of objects you allocate. This is not specific to destructuring synatx, but is instead a general concept in all garbage collected languages.

But I think our primary focus as developers should be to write clear, understandable code, that could be easily maintained later. So I suggest you to just use destructuring, and only optimize it when you actually hit a performance problem, and only after you measured that the performance can really be improved removing arguments destructuring.

> Even when we don't transpile and have native support (e.g. Node.js or Electron apps with definite support) - is this problem out of the world? Or do we just ditch it out of the Javascript layer and into the C layer below?
Of course it depends on the JS engine, but what do we know about e.g. V8 handling this?
Do we create much extra work for cleaning up our "garbage" inside V8? (I guess, while there is no GC in C++, there still needs to be some cleanup/allocation management)

Look at this graph, it show you the evolutions of performance difference between native es6 syntaxes and equivalent es5 ones:

![image](https://user-images.githubusercontent.com/11197111/38817981-d7d7f1ee-4199-11e8-85e5-35b5a3ff9d75.png)
> _image taken from https://v8project.blogspot.it/2017/02/high-performance-es2015-and-beyond.html_

The destructuring syntax lines are particularly interesting. In V8 5.4 native destructuring syntax is 10 times slower than the transpiled one. In V8 5.8 the speed is equivalent. And the V8 team is constantly improving on speed, so I won't worry to much about spreading syntax specifically.

And finally, regarding C++, you're right that you still have to manually collect memory you allocate for your object, and you still pay a performance penalty if you allocate a lot of them. But the great advantage is that you can get fine grain control on __when__ you collect your garbage. You can avoid doing it in delicate moments, when your user would be much concerned on time lags, and wait until the app is idle.

**Comment by timoxley** [comment]
@parro-it @loopmode This question was recently discussed on twitter by v8 team member @bmeurer: 

* https://twitter.com/bmeurer/status/987563203919048705
* https://twitter.com/bmeurer/status/987582050692628481

**Comment by loopmode** [comment]
Interesting, thanks! And of course, a good moment to recall the effects of poly/monomorphism on optimization.

**Comment by parro-it** [comment]
Thank you @timoxley, really interesting...
@loopmode, I prefer to keep questions opened. I'm still unsure on what to do if/when they will become a lot...

**Comment by SachinBhandari** [comment]
```
  var name = _ref.name,
  age = _ref.age,
  email = _ref.email;
```

> But this code only create new variables that refers of existing objects.

@parro-it What if, when these values `name`, `age`, `email` are primitive data types. Then they must not be holding the references anymore but instead creates new variables to hold these primitive data types.

What I think is, if we do not tend to change the value of these variables then it is good to access them as `x.name` but suppose if we want to have a copy of these variables and apply some filter like `name.trim()` then it is good to destructure it becasue if we don't have **destructuring** then anyway we are going to create a copy of these property.

**Comment by loopmode** [comment]
Well, in a mutable world, you can also do `x.name = x.name.trim()` without creating a new var, but that could create horrible side effects and I wouldn't recommend such practice :) also, your statement holds - as you explicitly spoke of cases when we need an actual copy of the value. Then we must allocate new memory (or: a new var) anyways.

However, coming back to this, my interest with the initial question wasn't the three primitive value vars, but the extra (and non-primitive) `_ref` object that surely becomes useless and gets GCed after the function ran.

**Comment by SachinBhandari** [comment]
@loopmode, Yeah modifying directly the property is a way and that's a bad idea, that's why I haven't put it. But that's not the concern. And about `_ref` that identifier is only holding the reference it gets when `createUser` is called with the actual object. So creating a new object costs you more that just passing it and `_ref` is GCed after the function ran

**Comment by loopmode** [comment]
Yes exactly! Which was my initial point - back in flash games, it was noticable in fps drops when GC occasionally kicked in and cleaned up a couple hundred thousand objects. (E.g event objects, `new Point(x,y)` etc). Which was why I started to use pooling techniques and reused objects.

Anyhow it's too theoretical here. I guess when I make a JS game, I'll prefer `updatePos(x, y)` over `updatePos({ x, y })` :). While I do webapps tho, I'll keep enjoying destructuring.

**Comment by SachinBhandari** [comment]
Yeah. One should look these kind of issues when they hit performance issues. 

> If you create lot of objects, you should always consider that garbage collection

I also wonder what about very deep objects. Do they cost performance issues?
